pkg/credentialstores/common.go

```diff
@@ -0,0 +1,129 @@
+package credentialstores
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/openshift/image-registry/pkg/kubernetes-common/credentialprovider"
+
+	"github.com/golang/glog"
+)
+
+// MultiError is a wrap for multiple errors.
+type MultiError struct {
+	errors []string
+}
+
+// Append adds a new error.
+func (m *MultiError) Append(e error) {
+	m.errors = append(m.errors, e.Error())
+}
+
+// Error exists to comply with golang's error interface.
+func (m *MultiError) Error() string {
+	return strings.Join(m.errors, " ")
+}
+
+// basicCredentialsFromKeyring extract basicAuth information from provided
+// keyring. If keyring does not contain information for the provided URL, empty
+// strings are returned instead.
+func basicCredentialsFromKeyring(keyring credentialprovider.DockerKeyring, target *url.URL) (string, string) {
```

Comment:
Why does it accept URL?

Reply:
Because docker/distribution uses it so makes sense to keep it concise:

https://github.com/docker/distribution/blob/84f47e7bb3bb3521d7d20615d42078df306fb01d/registry/client/auth/session.go#L44-L55

Reply:
DockerKeyring.Lookup accepts an image reference, not an URL. When a user provides you an image reference, you convert it into an URL, and then convert it back to an image reference-like string - you are doing something wrong.

CredentialStore is not supposed to be used that way, and we should never use it that way. You can find examples how to use it in the docker/distribution repostory.

Reply:
`docker/distribution` uses CredentialStores and k8s has something called keyring, the first uses url.URL and the latter string.

k8s makes available to us functions to parse secrets into keyrings and to use them we need these keyrings to be available on our project, hence we need a conversion layer that wraps a keyring with multiple credentials into a CredentialStore as needed by library-go registryclient.

Could you tell me what would you do differently? I would most certainly adapt this to your input as this codes predates this PR and myself.